yb-voyager/src/srcdb/status.go

```diff
@@ -0,0 +1,111 @@
+package srcdb
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/yugabyte/yb-voyager/yb-voyager/src/utils"
+)
+
+type TableExportStatus struct {
+	TableName                string `json:"table_name"` // qualified table name
+	FileName                 string `json:"file_name"`
+	Status                   string `json:"status"`
+	ExportedRowCountSnapshot int64  `json:"exported_row_count_snapshot"`
+}
+
+type ExportSnapshotStatus struct {
+	Tables     map[string]*TableExportStatus `json:"tables"`
+	statusLock sync.Mutex                    `json:"-"` // not serialized
+	FilePath   string                        `json:"-"` // not serialized
+}
+
+func NewExportSnapshotStatus(exportDir string) *ExportSnapshotStatus {
+	return &ExportSnapshotStatus{
+		Tables:   make(map[string]*TableExportStatus),
+		FilePath: filepath.Join(exportDir, "metainfo", "export_snapshot_status.json"),
+	}
+}
+
+func ReadExportStatus(statusFilePath string) (*ExportSnapshotStatus, error) {
+	file, err := os.Open(statusFilePath)
```

Comment:
Consider implementing a utility class called `JsonFile` as:
src/utils/jsonfile/jsonfile.go
```
package jsonfile

import (
	"encoding/json"
	"fmt"
	"os"
	"sync"

	"github.com/yugabyte/yb-voyager/yb-voyager/src/utils"
)

type JsonFile[T any] struct {
	sync.Mutex
	FilePath string
}

func NewJsonFile[T any](filePath string) *JsonFile[T] {
	return &JsonFile[T]{FilePath: filePath}
}

func (j *JsonFile[T]) Read() (*T, error) {
	j.Lock()
	defer j.Unlock()
	return j.read()
}

func (j *JsonFile[T]) read() (*T, error) {
	bs, err := os.ReadFile(j.FilePath)
	if err != nil {
		return nil, fmt.Errorf("read file %s: %w", j.FilePath, err)
	}
	if len(bs) == 0 {
		return nil, fmt.Errorf("file %s is empty", j.FilePath)
	}
	obj := new(T)
	err = json.Unmarshal(bs, obj)
	if err != nil {
		return nil, fmt.Errorf("unmarshal json: %w", err)
	}
	return obj, nil
}

func (j *JsonFile[T]) Update(fn func(*T)) error {
	j.Lock()
	defer j.Unlock()
	var obj *T
	var err error
	if utils.FileOrFolderExists(j.FilePath) {
		obj, err = j.read()
		if err != nil {
			return err
		}
	} else {
		obj = new(T)
	}
	fn(obj)
	return j.write(obj)
}

func (j *JsonFile[T]) write(obj *T) error {
	bs, err := json.MarshalIndent(obj, "", "  ")
	if err != nil {
		return fmt.Errorf("marshal json: %w", err)
	}
	err = os.WriteFile(j.FilePath, bs, 0644)
	if err != nil {
		return fmt.Errorf("write file %s: %w", j.FilePath, err)
	}
	return nil
}

func (j *JsonFile[T]) Delete() error {
	j.Lock()
	defer j.Unlock()
	return os.Remove(j.FilePath)
}
```
And `jsonfile_test.go`
```
package jsonfile

import (
	"os"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestJsonFile(t *testing.T) {
	type Person struct {
		Name string
	}
	var person *Person
	jf := NewJsonFile[Person]("person.json")
	_ = jf.Delete()
	person, err := jf.Read()
	assert.ErrorIs(t, err, os.ErrNotExist)
	assert.Nil(t, person)
	err = jf.Update(func(p *Person) {
		p.Name = "John Doe"
	})
	assert.Nil(t, err)
	person, err = jf.Read()
	assert.Nil(t, err)
	assert.NotNil(t, person)
	assert.Equal(t, "John Doe", person.Name)
	err = jf.Update(func(p *Person) {
		p.Name = "John Smith"
	})
	assert.Nil(t, err)
	person, err = jf.Read()
	assert.Nil(t, err)
	assert.Equal(t, "John Smith", person.Name)
	err = jf.Delete()
	assert.Nil(t, err)
}
```
Once you have the `JsonFile` implementation:
- you can make the `ExportSnapshotStatus` a plain data object without any methods.
- you can directly define and use `exportSnapshotStatusFile JsonFile[ExportSnapshotStatus]`.